lib/burnside/mixin.h

```diff
@@ -0,0 +1,133 @@
+
+namespace Fortran::burnside {
+
+// implementation of a (moveable) sum type (variant)
+template<typename... Ts> struct SumTypeMixin {
+  using SumTypeTrait = std::true_type;
+  template<typename A> SumTypeMixin(const A &x) : u{x} {}
+  template<typename A> SumTypeMixin(A &&x) : u{std::forward<A>(x)} {}
+  SumTypeMixin(SumTypeMixin &&) = default;
+  SumTypeMixin &operator=(SumTypeMixin &&) = default;
+  SumTypeMixin(const SumTypeMixin &) = delete;
+  SumTypeMixin &operator=(const SumTypeMixin &) = delete;
+  SumTypeMixin() = delete;
+  std::variant<Ts...> u;
+};
+
+// implementation of a copyable sum type
+template<typename... Ts> struct SumTypeCopyMixin {
+  using CopyableSumTypeTrait = std::true_type;
+  template<typename A> SumTypeCopyMixin(const A &x) : u{x} {}
+  template<typename A> SumTypeCopyMixin(A &&x) : u{std::forward<A>(x)} {}
+  SumTypeCopyMixin(SumTypeCopyMixin &&) = default;
+  SumTypeCopyMixin &operator=(SumTypeCopyMixin &&) = default;
+  SumTypeCopyMixin(const SumTypeCopyMixin &) = default;
+  SumTypeCopyMixin &operator=(const SumTypeCopyMixin &) = default;
+  SumTypeCopyMixin() = delete;
+  std::variant<Ts...> u;
+};
+#define SUM_TYPE_COPY_MIXIN(DT) \
+  DT(const DT &derived) : SumTypeCopyMixin(derived.u) {} \
+  DT(DT &&derived) : SumTypeCopyMixin(std::move(derived.u)) {} \
+  DT &operator=(const DT &derived) { \
+    SumTypeCopyMixin::operator=(derived.u); \
+    return *this; \
+  } \
+  DT &operator=(DT &&derived) { \
+    SumTypeCopyMixin::operator=(std::move(derived.u)); \
+    return *this; \
+  }
+
+// implementation of a (moveable) product type (tuple)
+template<typename... Ts> struct ProductTypeMixin {
+  using ProductTypeTrait = std::true_type;
+  ProductTypeMixin(const Ts &... x) : t{x...} {}
+  template<typename... As>
+  ProductTypeMixin(As &&... x) : t{std::forward<As>(x)...} {}
+  ProductTypeMixin(ProductTypeMixin &&) = default;
+  ProductTypeMixin &operator=(ProductTypeMixin &&) = default;
+  ProductTypeMixin(const ProductTypeMixin &) = delete;
+  ProductTypeMixin &operator=(const ProductTypeMixin &) = delete;
+  ProductTypeMixin() = delete;
+  std::tuple<Ts...> t;
+};
+
+// implementation of a (moveable) maybe type
+template<typename T> struct MaybeMixin {
+  using MaybeTrait = std::true_type;
+  MaybeMixin(const T &x) : o{x} {}
+  MaybeMixin(T &&x) : o{std::move(x)} {}
+  MaybeMixin(MaybeMixin &&) = default;
+  MaybeMixin &operator=(MaybeMixin &&) = default;
+  MaybeMixin(const MaybeMixin &) = delete;
+  MaybeMixin &operator=(const MaybeMixin &) = delete;
+  MaybeMixin() = delete;
+  std::optional<T> o;
+};
+
+// implementation of a child type (composable hierarchy)
+template<typename T, typename P> struct ChildMixin {
+protected:
+  P *parent;
+
+public:
+  ChildMixin(P *p) : parent{p} {}
+  inline const P *getParent() const { return parent; }
+  inline P *getParent() { return parent; }
+  llvm::iplist<T> &getList() { return parent->getSublist(this); }
+};
+
+// zip :: ([a],[b]) -> [(a,b)]
```

Comment:
These three template functions are not used; if they were, `lib/common` might be a better place for them as general utilities.